src/connectors/impls/gcs/api_client.rs

```diff
@@ -0,0 +1,476 @@
+
+            match self.client.send(request).await {
+                Ok(request) => response = Some(request),
+                Err(e) => {
+                    warn!("Failed to send a request to GCS: {}", e);
+
+                    sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+                    continue;
+                }
+            }
+
+            if let Some(response) = response.as_mut() {
+                if !response.status().is_server_error() && response.header("range").is_some() {
+                    break;
+                }
+
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+
+                sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+            }
+        }
+
+        if let Some(mut response) = response {
+            if response.status().is_server_error() {
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+                return Err("Received server errors from Google Cloud Storage".into());
+            }
+
+            if let Some(raw_range) = response.header("range") {
+                let raw_range = raw_range
+                    .get(0)
+                    .ok_or(ErrorKind::GoogleCloudStorageError(
+                        "Missing Range header value",
+                    ))?
+                    .as_str();
+
+                // Range format: bytes=0-262143
+                let range_end = &raw_range
+                    .get(
+                        raw_range
+                            .find('-')
+                            .ok_or(ErrorKind::GoogleCloudStorageError(
+                                "Did not find a - in the Range header",
+                            ))?
+                            + 1..,
+                    )
+                    .ok_or(ErrorKind::GoogleCloudStorageError(
+                        "Unable to get the end of the Range",
+                    ))?;
+
+                // NOTE: The data can only be thrown away to the point of the end of the Range header,
+                // since google can persist less than we send in our request.
+                //
+                // see: https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+                done_until.store(range_end.parse()?, Ordering::Release);
+            } else {
+                return Err("No range header".into());
+            }
+        } else {
+            return Err("no response from GCS".into());
+        }
+
+        Ok(())
+    }
+
+    async fn start_upload(&mut self, url: &Url<HttpsDefaults>, file: FileId) -> Result<()> {
+        for i in 0..3 {
+            let url = url::Url::parse(&format!(
+                "{}/b/{}/o?name={}&uploadType=resumable",
+                url, file.bucket, file.name
+            ))?;
+            #[cfg(not(test))]
```

Comment:
can we isolate the `test` and `non test` code into a separate function that's swapped out? mixing `cfg(test)` into the code makes it extremely hard to follow.